api/include/opentelemetry/logs/logger.h

```diff
@@ -0,0 +1,138 @@
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace logs
+{
+/**
+ * Handles log record creation.
+ **/
+class Logger
+{
+public:
+  virtual ~Logger() = default;
+
+  /* Returns the name of the logger */
+  // TODO: decide whether this is useful and/or should be kept, as this is not a method required in
+  // the specification. virtual nostd::string_view getName() = 0;
+
+  /**
+   * Each of the following overloaded log(...) methods
+   * creates a log message with the specific parameters passed.
+   *
+   * @param name the name of the log event.
+   * @param severity the severity level of the log event.
+   * @param message the string message of the log (perhaps support std::fmt or fmt-lib format).
+   * @param record the log record (object type LogRecord) that is logged.
+   * @param attributes the attributes, stored as a 2D list of key/value pairs, that are associated
+   * with this log.
+   * @throws No exceptions under any circumstances.
+   */
+
+  /* The below method is a logging statement that takes in a LogRecord.
+   * A default LogRecord that will be assigned if no parameters are passed to Logger's .log() method
+   * which should at minimum assign the trace_id, span_id, and timestamp
+   */
+  virtual void log(const LogRecord &record) noexcept = 0;
+
+  /** Overloaded methods for unstructured logging **/
+  void log(nostd::string_view message) noexcept
+  {
+    // TODO: move this log method to the SDK? since it needs to call an SDK method
+    // "isEnabled(Severity)"
+
+    // if(isEnabled(Severity::kDefault)){
+    //   return;
+    // }
+
+    // Set severity to the default then call log(Severity, String message) method
+    log(Severity::kDefault, message);
+  }
+
+  void log(Severity severity, nostd::string_view message) noexcept
+  {
+    // TODO: move this log method to the SDK? since it needs to call an SDK method
+    // "isEnabled(Severity)" if(isEnabled(Severity::kDefault)){
+    //   return;
+    // }
+
+    // TODO: set default timestamp later (not in API)
+    log(severity, message, core::SystemTimestamp(std::chrono::system_clock::now()));
+  }
+
+  void log(Severity severity, nostd::string_view message, core::SystemTimestamp time) noexcept
+  {
+    // TODO: move this log method to the SDK? since it needs to call an SDK method
+    // "isEnabled(Severity)" if(isEnabled(Severity::kDefault)){
+    //   return;
+    // }
+
+    // creates a LogRecord object with given parameters, then calls log(LogRecord)
+    LogRecord r;
+    r.severity  = severity;
+    r.body      = message;
+    r.timestamp = time;
+
+    log(r);
+  }
+
+  /** Overloaded methods for structured logging**/
+  // TODO: separate this method into separate methods since it is not useful for user to create
+  // empty logs
+  template <class T,
+            nostd::enable_if_t<common::detail::is_key_value_iterable<T>::value> * = nullptr>
```

Comment:
I think you can also safely add a template for `log(severity, name, T)`, where T is NOT `KeyValueIterable`. This would live in API header, not implemented, but allows a customer to provide their own implementation. That way if I do `log(severity, name,  nlohmann::json obj)` - I can plug in my own `flattener` implementation outside of SDK to existing class. Not in this PR, but consider this. I can help to provide this as a `contrib` example to show how to use this feature.. We can theoretically support wider set of types, and have that outside-of-SDK implementation to flatten into OT typing system. This will certainly be needed to migrate customers that might have been using wider typing system in their other logging framework.

Reply:
Yea I totally agree with this, we'll submit this addition in a later commit! We'll leave it as a non-pure virtual function so that a definition isn't required to compile.